mdal/mdal_memory_data_model.hpp

```diff
@@ -0,0 +1,98 @@
+/*
+ MDAL - Mesh Data Abstraction Library (MIT License)
+ Copyright (C) 2018 Peter Petrik (zilolv at gmail dot com)
+*/
+
+#ifndef MDAL_MEMORY_DATA_MODEL_HPP
+#define MDAL_MEMORY_DATA_MODEL_HPP
+
+#include <stddef.h>
+#include <vector>
+#include <memory>
+#include <map>
+#include <string>
+#include "mdal.h"
+#include "mdal_data_model.hpp"
+
+namespace MDAL
+{
+  typedef struct
+  {
+    double x;
+    double y;
+    double z; // Bed elevation
+  } Vertex;
+
+  typedef std::vector<size_t> Face;
+  typedef std::vector<Vertex> Vertices;
+  typedef std::vector<Face> Faces;
+
+  /**
+   * The MemoryDataset stores all the data in the memory
+   */
+  class MemoryDataset: public Dataset
+  {
+    public:
+      ~MemoryDataset() override;
+
+      size_t scalarData( size_t indexStart, size_t count, double *buffer ) override;
+      size_t vectorData( size_t indexStart, size_t count, double *buffer ) override;
+      size_t activeData( size_t indexStart, size_t count, int *buffer ) override;
+
+      /**
+       * size - face count if !isOnVertices
+       * size - vertex count if isOnVertices
+       */
+      std::vector<Value> values;
```

Comment:
I would suggest to have two different arrays - one for scalars and one for vectors. Like this we are wasting a lot of memory.

Also, noData in Value seems unneeded and bloats the structure (8 + 8 + 4 bytes + 4 bytes alignment (I think)). How about:
```
std::vector<double> valuesScalar;
struct double2d { double x, double y };
std::vector<double2d> valuesVector;
```
This would even allow us to use simple memcpy to fill up the buffers.

Reply:
@wonder-sk or something like this?
```
  class Values {
    public:
      Values(size_t nItems, bool isVector);

      void setX(size_t index, double val);
      void setY(size_t index, double val); // only for isVector = true

      double x(size_t index) const;
      double y(size_t index) const; // only for isVector = true

      double* data();
      double* constData() const;
    private:
      bool mIsVector;
      // nItems * 2 for vectors
      // nItems for scalars
      // x=NaN, y=NaN for NODATA values
      std::vector<double> mValues;
  };
```

Reply:
Hmm... I don't like the idea to have a class with X/Y getters/setters when that only makes sense for vectors. I would rather go with just simple `std::vector<double> value` then. Or to have two separate classes - e.g. ScalarValues / Vector2DValues if we want full classes (but we probably don't need that).